### DISCUSSION TITLE
server doesn't seem to be starting

### DISCUSSION THREAD

**Discussion by nminchow** [mbround18/valheim-docker#115] [Help!] [answered]
I'm attempting to spin this up without compose on a dedicated vm, and I'm not sure where I should continue debugging. The container starts and runs, but its status is "unhealthy" and I can't connect to the server.

`docker ps`:
```
  CREATED             STATUS                     PORTS               NAMES
dad2c1fa27ed        mbround18/valheim:latest                                             "/bin/bash /entrypoi…"   10 minutes ago      Up 9 minutes (unhealthy)                       klt-valheim-odin-yrps
```

Initially, I didn't supply `PUID` or `PGID`, but have since switched them to my typical user on the host machine's UID.

logs:
```
[Valheim][root] : ###########################################################################
[Valheim][root] : Valheim Server - Tue Feb 16 10:04:07 PST 2021
[Valheim][root] : Initializing your container...
[Valheim][root] : ###########################################################################
[Valheim][root] : Switching UID and GID
[Valheim][root] : 
[Valheim][root] : 
[Valheim][root] : Auto Update Enabled...
[Valheim][root] : Auto Update Schedule: 0 1 * * *
rm: cannot remove '/home/steam/valheim/logs/auto-update.out': No such file or directory
[Valheim][root] : Auto Backup Enabled...
[Valheim][root] : Auto Backup Schedule: 0 * * * *
rm: cannot remove '/home/steam/valheim/logs/auto-backup.out': No such file or directory
[Valheim][root] : Setting up file systems
[Valheim][root] : Launching as steam...
[Valheim][steam]: ###########################################################################
[Valheim][steam]: Valheim Server - Tue Feb 16 10:04:09 PST 2021
[Valheim][steam]: STEAM_UID 20141 - STEAM_GUID 20141
[Valheim][steam]: Installing Valheim via Odin...
[Valheim][steam]: ###########################################################################
[Valheim][steam]: Variables loaded.....
[Valheim][steam]: 
Port: 2456
Name: Valgrind
World: Valgrind
Public: 1
Password: (REDACTED)

[Valheim][steam]: Running Install...
[ODIN][INFO]  - Installing 896660 to /home/steam/valheim
[ODIN][ERROR] - Failed to find steamcmd in path
[ODIN][ERROR] - Checking for script under steam user.
[ODIN][INFO]  - Executing: /home/steam/steamcmd/steamcmd.sh .....
[ODIN][INFO]  - Using steamcmd script at /home/steam/steamcmd/steamcmd.sh
WARNING: setlocale('en_US.UTF-8') failed, using locale: 'C'. International characters may not work.
Redirecting stderr to '/home/steam/Steam/logs/stderr.txt'
/tmp/dumps is not owned by us - delete and recreate
Unable to delete /tmp/dumps. Continuing anyway.
[  0%] Checking for available updates...
[----] Downloading update (139 of 16258 KB)...
[  0%] Downloading update (2633 of 16258 KB)...
[ 16%] Downloading update (4307 of 16258 KB)...
[ 26%] Downloading update (5737 of 16258 KB)...
[ 35%] Downloading update (7031 of 16258 KB)...
[ 43%] Downloading update (8295 of 16258 KB)...
[ 51%] Downloading update (9544 of 16258 KB)...
[ 58%] Downloading update (10799 of 16258 KB)...
[ 66%] Downloading update (12184 of 16258 KB)...
[ 74%] Downloading update (13431 of 16258 KB)...
[ 82%] Downloading update (14699 of 16258 KB)...
[ 90%] Downloading update (15961 of 16258 KB)...
[ 98%] Downloading update (16258 of 16258 KB)...
[100%] Download complete.
[----] Installing update...
[----] Extracting package...
[----] Extracting package...
[----] Extracting package...
[----] Extracting package...
[----] Installing update...
[----] Installing update...
[----] Installing update...
[----] Installing update...
[----] Installing update...
[----] Installing update...
[----] Installing update...
[----] Installing update...
[----] Cleaning up...
[----] Update complete, launching Steamcmd...
WARNING: setlocale('en_US.UTF-8') failed, using locale: 'C'. International characters may not work.
Redirecting stderr to '/home/steam/Steam/logs/stderr.txt'
/tmp/dumps is not owned by us - delete and recreate
Unable to delete /tmp/dumps. Continuing anyway.
[  0%] Checking for available updates...
[----] Verifying installation...
Steam Console Client (c) Valve Corporation
-- type 'quit' to exit --
Loading Steam API...OK.

Connecting anonymously to Steam Public...Logged in OK
Waiting for user info...OK
Success! App '896660' already up to date.
[ODIN][INFO]  - Successfully installed Valheim!
[Valheim][steam]: Initializing Variables....
[ODIN][INFO]  - Successfully written config.json
[Valheim][steam]: Herding Cats...
[Valheim][steam]: Starting server...
[ODIN][INFO]  - Setting up start scripts...
[ODIN][INFO]  - Looking for burial mounds...
[ODIN][INFO]  - Server has been started and Daemonized. It should be online shortly!
==> /home/steam/valheim/logs/valheim_server.log <==
[ODIN][INFO]  - Executing: /home/steam/valheim/valheim_server.x86_64 .....
[ODIN][INFO]  - Success, daemonized

==> /home/steam/valheim/logs/valheim_server.err <==

==> /home/steam/valheim/logs/output.log <==
[Valheim][steam]: ###########################################################################
[Valheim][steam]: Valheim Server - Tue Feb 16 10:04:26 PST 2021
[Valheim][steam]: STEAM_UID 20141 - STEAM_GUID 20141
[Valheim][steam]: 
Valheim Server Started...

Keep an eye out for 'Game server connected' in the log!
(this indicates its online without any errors.)

[Valheim][steam]: ###########################################################################

==> /home/steam/valheim/logs/auto-update.out <==

==> /home/steam/valheim/logs/auto-backup.out <==

```

**Comment by CosmicHorrorDev** [comment]
Very common problem first:

How much RAM are you giving the VM? The server by itself will take about 2GB, so the VM would likely need 4GB total. Unhealthy means that the server process is no longer running.

**Reply by nminchow** [reply]
Yup - 4GB, and this is the only thing on the box.

**Reply by CosmicHorrorDev** [reply]
Hmm, could you get an interactive shell under the steam user

```sh
docker exec --user steam -it <the-container-name> bash
```

And try running `odin start --dry-run` and then copy and run the command manually to see if that actually starts the server? Note doing it this way you'll need to escape some special characters in any of the args so I would just stick to alphanumeric for debugging

**Reply by nminchow** [reply]
sure thing:
```
steam@valheim-odin:~/valheim$ odin start --dry-run
[ODIN][INFO]  - Setting up start scripts...
[ODIN][INFO]  - Looking for burial mounds...
[ODIN][INFO]  - This command would have launched
[ODIN][INFO]  - /home/steam/valheim/valheim_server.x86_64 -port 2456 -name Valgrind -world Valgrind -password 12345 -public 1
```

**Reply by CosmicHorrorDev** [reply]
Woops `odin --dry-run start`, not the above

**Reply by nminchow** [reply]
looks like the same output:

```
steam@valheim-odin:~/valheim$ odin --dry-run start
[ODIN][INFO]  - Setting up start scripts...
[ODIN][INFO]  - Looking for burial mounds...
[ODIN][INFO]  - This command would have launched
[ODIN][INFO]  - /home/steam/valheim/valheim_server.x86_64 -port 2456 -name Valgrind -world Valgrind -password 12345 -public 1
```

**Reply by CosmicHorrorDev** [reply]
Nothing looks too off there either. Can you try running that command from the interactive shell to see if the server launches.

Also just so I can reproduce, what OS are you using for the VM, how many resources (CPU, RAM, disk) are allocated, and what did you use to setup the docker server (docker compose, just docker, something else)?

**Reply by nminchow** [reply]
If I run it normally (`odin start`) it doesn't look like `output.log` has any changes. `valheim_server.err` does (its empty) and `valheim_server.log` is updated, but it just contains:
```
[ODIN][INFO]  - Executing: /home/steam/valheim/valheim_server.x86_64 .....
[ODIN][INFO]  - Success, daemonized
```

**Reply by CosmicHorrorDev** [reply]
My bad, I mean the command listed by in dry run. The `valheim_server.x86_64 ...` one

**Reply by nminchow** [reply]
Ah, interesting, I get a `permission denied`:
```
steam@valheim-odin:~/valheim$ /home/steam/valheim/valheim_server.x86_64 -port 2456 -name Valgrind -world Valgrind -password 12345 -public 1
bash: /home/steam/valheim/valheim_server.x86_64: Permission denied
```
Again, I am passing in my host user's PUID or PGID (which I wasn't doing initially)

```
steam@valheim-odin:~/valheim$ ls -l /home/steam/valheim/valheim_server.x86_64
-rwxr-xr-x 1 steam steam 6344 Feb 16 09:27 /home/steam/valheim/valheim_server.x86_64
```


I'm on google's [container optimized OS](https://cloud.google.com/container-optimized-os/docs) (my first time using it). 2vCPU, 4GB of ram.

**Reply by mbround18** [reply]
If you were not passing in your PUID and PGID originally, try deleting the server installation under `/home/steam/valheim`  and relaunching the container.

**Reply by mbround18** [reply]
Holy crap, 25 a month for shared resources???

**Reply by nminchow** [reply]
Ok - I was able to delete everything other than the folder itself (I presume due to how it was mounted as a volume). I `777`ed the directory and am now relaunching.

> Holy crap, 25 a month for shared resources???

I have a bunch of projects in gcp, so this is just easier to manage lol. I also have built some tooling which spins the VM down when it isn't in use and spins it up on-demand, so it should come out to a few bucks a month.

**Reply by nminchow** [reply]
> If you were not passing in your PUID and PGID originally, try deleting the server installation under /home/steam/valheim and relaunching the container.

No dice. Still "unhealthy" with nothing in the logs.

**Reply by mbround18** [reply]
I launched it as well in googles cloud, how are you passing in the PUID and PGID?

**Reply by nminchow** [reply]
> I launched it as well in googles cloud, how are you passing in the PUID and PGID?

Oh, wow - I appreciate the help debugging!

"instance details" > "advanced container options" > "Environment variables "

**Reply by nminchow** [reply]
I am also mounting under the default user's "home" directory, but I wouldn't think that would matter:

```
Mount path 	Host path 	Mode
/home/steam/.config/unity3d/IronGate/Valheim 	/home/myusername/valheim/saves 	Read/write
/home/steam/valheim 	/home/myusername/valheim/server 	Read/write
/home/steam/backups 	/home/myusername/valheim/backups 	Read/write
```

**Reply by mbround18** [reply]
it does lol @nminchow I got it to work, remove the PGID and PUID variables :P Then make sure the mount points are in `/home/your-google-username/some-dir`

Additionally, to make your mount points simple and unless you want to install mods. You dont need to mount the server directory :)  

![image](https://user-images.githubusercontent.com/12646562/108116218-165fbe00-7050-11eb-84e0-8aa6f890e270.png)

**Reply by nminchow** [reply]
yeah - they are under my user directory. I'll try without mounting the server directory and remove those env variables.

**Reply by nminchow** [reply]
Unmounting the server directory seemed to do the trick! Thanks again for the help! Any guesses as to why that would matter?

**Reply by mbround18** [reply]
Permissions on the cloud container vs cloud os and their security on the inner workings of the container. It automatically forces mounted directories to be non-executable on the host OS which is why it couldnt execute the server executable.

**Reply by nminchow** [reply]
ah - got it. I don't think it will matter for me anyway. If you want to summarize that as a top response, I'd be happy to mark it as an answer for posterity! Thanks again.

**Comment by mbround18** [comment, marked as answer]
Permissions on the cloud container vs cloud os and their security on the inner workings of the container. It automatically forces mounted directories to be non-executable on the host OS which is why it couldnt execute the server executable.

Thus running inside a Google container optimized OS requires you to not mount the server directory. You can skip that mount since it is an optional mount.
